### PR TITLE
feat: add l2-bridge-base

### PR DESCRIPTION
**Pull request by sergeyWh1te**

### CODE REVIEW

**Review by arwer13 — CHANGES_REQUESTED** [review]
Thank you, looks fine to me 👍. I've left only a few minor refactor-related comments.

In general I think we should getting out of code duplication: I mean arbitrum, optimism and base and there are more L2s to come. But I propose it to be one of the next iterations.

File: l2-bridge-base/src/agent.ts [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+
+import * as agentGov from "./agent-governance";
+import * as agentProxy from "./agent-proxy-watcher";
+import * as agentBridge from "./agent-bridge-watcher";
+import * as agentWithdrawals from "./agent-withdrawals";
+import * as agentBalance from "./agent-balance";
+import VERSION from "./version";
+
+type Metadata = { [key: string]: string };
+type CustomHandleBlock = (blockEvent: BlockDto) => Promise<Finding[]>;
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply l2-bridge-base/src/agent.ts:25]: dto - it's a common pattern to name objects. Dto (data transfer object)
For example: https://www.okta.com/identity-101/dto/

File: l2-bridge-base/src/agent.ts [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+const timeout = async (agent: SubAgent) =>
+  new Promise<never>((_, reject) => {
+    setTimeout(() => {
+      const err = new Error(`Sub-agent ${agent.name} timed out`);
+      reject(err);
+    }, processingTimeout);
+  });
+
+let cachedBlockDto: BlockDto;
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply l2-bridge-base/src/agent.ts:123]: Delegate this question to forta team.

File: l2-bridge-base/src/agent-withdrawals.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    );
+    lastReportedToManyWithdrawals = now;
+    // remove reported withdrawals records
+    withdrawalsCache = withdrawalsCache.filter(
+      (x: IWithdrawalRecord) => x.time > lastReportedToManyWithdrawals,
+    );
+  }
+}
+
+export async function handleTransaction(logs: Log[], blockEvents: BlockDto[]) {
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply l2-bridge-base/src/agent-withdrawals.ts:123]: fixed

File: l2-bridge-base/src/agent.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (arwer13) [reply l2-bridge-base/src/agent.ts:25]: Got it

File: l2-bridge-base/src/agent-balance.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+export const name = "Balance";
+
+export async function initialize(
+  currentBlock: number,
+): Promise<{ [key: string]: string }> {
+  console.log(`[${name}]`);
+  return {};
+}
+
+export async function handleBlock(block: BlockDto) {
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply l2-bridge-base/src/agent-balance.ts]: moved

File: l2-bridge-base/src/agent.ts [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+    workingBlocks.push(cachedBlockDto);
+    iteration += 1;
+  } else {
+    const latestBlock: Block = await baseProvider.getBlock("latest");
+    const range = function (start: number, end: number): number[] {
+      return Array.from(Array(end - start + 1).keys()).map((x) => x + start);
+    };
+
+    const blocksInterval = range(cachedBlockDto.number, latestBlock.number);
```
[the remark this answers is not in the document]
Comment (arwer13) [reply l2-bridge-base/src/agent.ts:147]: Fixed

**Review by arwer13 — DISMISSED** [review]
I'm also a little concerned about the lack of tests e.g. like e2e in ethereum-steth agent, but as far as it lacked in optimism, let's consider the tests in the next iteration.

**Review by arwer13 — APPROVED** [review]

**Review by arwer13 — APPROVED** [review]

**Review by dgusakov — DISMISSED** [review]

**Review by arwer13 — APPROVED** [review]

**Review by dgusakov — APPROVED** [review]

**Pull request merged**